src/hypothesis/internal/conjecture/data.py

```diff
@@ -160,7 +160,14 @@ def stop_example(self, discard=False):
             t = (k, self.index)
             self.intervals_by_level[self.level].append(t)
             if discard:
-                self.discarded.add(t)
+                while self.discarded:
+                    u, v = t
+                    r, s = self.discarded[-1]
+                    if u <= r <= s <= v:
+                        self.discarded.pop()
+                    else:
+                        break
+                self.discarded.append(t)
```

Comment:
I'd finish with `assert self.discarded == sorted(self.discarded)`, just to guard against weird changes in the interval stack.

Reply:
I wouldn't want to do it here (it brings us back to quadratic!), and we sort at point of use anyway, so I'm not sure this is needed.

Reply:
Ah, there's a cheaper assertion we can add that does the trick: `assert s <= u`. I'll add that.